src/operations.c

```diff
@@ -25,14 +25,49 @@
 #include "baton.h"
 #include "operations.h"
 
+// The connection used by iterate_json and connection_timeout
+rcComm_t *connection;
+// Mutex protecting the connection
+pthread_mutex_t conn_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+// While true, the client will refresh the connection
+int connection_timeout_flag = 1;
+
+// Refresh the connection every timeout seconds
+void *connection_timeout(void *timeout) {
+    int tsec = *((int *) timeout);
+
+    struct timespec t;
+    t.tv_sec = tsec;
+    t.tv_nsec = 0;
+
+    while (connection_timeout_flag) {
+        nanosleep(&t, NULL);
+
+        logmsg(DEBUG, "Connection timeout of %d seconds reached", tsec);
+        pthread_mutex_lock(&conn_mutex);
+        if (connection) {
+            rcDisconnect(connection);
+            connection = NULL;
+            logmsg(DEBUG, "Closed the connection after timeout reached");
+        }
+        pthread_mutex_unlock(&conn_mutex);
+    }
```

Comment:
You need to protect `connection_timeout_flag` with a mutex, or possibly a conditional, otherwise you may be using a stale copy of the value.  Even worse, as nothing changes it inside this function, it could be copied into a register, so it will never change in this thread.

One way to do this might be to use `conn_mutex` to protect it:
```suggestion
    pthread_mutex_lock(&conn_mutex);
    while (connection_timeout_flag) {
        pthread_mutex_unlock(&conn_mutex);
        nanosleep(&t, NULL);

        logmsg(DEBUG, "Connection timeout of %d seconds reached", tsec);
        pthread_mutex_lock(&conn_mutex);
        if (connection) {
            rcDisconnect(connection);
            connection = NULL;
            logmsg(DEBUG, "Closed the connection after timeout reached");
        }
    }
    pthread_mutex_unlock(&conn_mutex);
```

Setting `connection_timeout_flag` to zero would also need to be protected via a mutex.  Initially setting it to 1 wouldn't, because that happens before your program starts.

However, another problem with this is that you have to wait for the timeout to occur before the thread eventually exits.  A better solution may be to use a conditional and `pthread_cond_timedwait()` instead.  Here you create a conditional variable at file scope:

```
pthread_cond_t watchdog_cond = PTHREAD_COND_INITIALIZER;
```

Then here you use it in place of the `nanosleep()`:
```suggestion
    struct timespec abs_timeout;
    pthread_mutex_lock(&conn_mutex);
    while (connection_timeout_flag) {
        int res;
        clock_gettime(CLOCK_REALTIME, &abs_timeout);
        abs_timeout.tv_sec += tsec;
        do {
            res = pthread_cond_timedwait(&watchdog_cond, &conn_mutex, &abs_timeout);
        } while (res == EINTR);
        if (res == ETIMEDOUT) {
            logmsg(DEBUG, "Connection timeout of %d seconds reached", tsec);
            if (connection) {
                rcDisconnect(connection);
                connection = NULL;
                logmsg(DEBUG, "Closed the connection after timeout reached");
            }
        }
    }
    pthread_mutex_unlock(&conn_mutex);
```

When clearing `connection_timeout_flag`, you would need to lock the mutex and signal on the condiiton, which will make the loop in the thread terminate as soon as it next runs:

```
    pthread_mutex_lock(&conn_mutex);
    connection_timeout_flag = 0;
    pthread_cond_signal(&watchdog_cond);
    pthread_mutex_unlock(&conn_mutex);
```

With the code above, you could also do an occasional `pthread_cond_signal(&watchdog_cond);` in your main loop while `connection_timeout_flag` is still 1 to reset the timeout without dropping the connection.

Reply:
Also note that while I've written the above code fragments, I haven't tested any of them...

Reply:
Thanks - I missed protecting the flag. I was aware that I'd have to wait for the thread to timeout - it was a choice to avoid the extra complexity of the condition. I'll do some testing based on your suggestions and will certainly fix the former.

Reply:
I fixed the flag protection and added your suggestion of a condition for timeout/thread exit.